src/extra/gfx/view.cpp

```diff
@@ -0,0 +1,157 @@
+    _viewTransform.warmup(context);
+  }
+
+  SHVar activate(SHContext *context, const SHVar &input) {
+    ViewPtr view = _view->view = std::make_shared<View>();
+    if (_viewTransform->valueType != SHType::None) {
+      view->view = shards::Mat4(_viewTransform.get());
+    }
+
+    // TODO: Add projection/viewport override params
+    view->proj = ViewPerspectiveProjection{};
+
+    if (_viewTransform.isVariable()) {
+      _view->viewTransformVar = (SHVar &)_viewTransform;
+      _view->viewTransformVar.warmup(context);
+    }
+    return Types::ViewObjectVar.Get(_view);
+  }
+};
+
+struct RegionShard : public BaseConsumer {
+  static SHTypesInfo inputTypes() { return CoreInfo::AnyTableType; }
+  static SHTypesInfo outputTypes() { return CoreInfo::AnyType; }
+  static SHOptionalString help() { return SHCCSTR("Renders within a region of the screen and/or to a render target"); }
+
+  PARAM(ShardsVar, _content, "Content", "The code to run inside this region", {CoreInfo::ShardsOrNone});
+  PARAM_IMPL(RegionShard, PARAM_IMPL_FOR(_content));
+
+  void cleanup() {
+    PARAM_CLEANUP();
+    baseConsumerCleanup();
+  }
+
+  void warmup(SHContext *context) {
+    baseConsumerWarmup(context);
+    PARAM_WARMUP(context);
+  }
+
+  SHTypeInfo compose(SHInstanceData &data) { return _content.compose(data).outputType; }
+
+  SHVar activate(SHContext *shContext, const SHVar &input) {
+    ViewStack::Item viewItem{};
+    input::InputStack::Item inputItem;
+
+    // (optional) set viewport rectangle
+    auto &inputTable = input.payload.tableValue;
+    SHVar viewportVar{};
+    if (getFromTable(shContext, inputTable, "Viewport", viewportVar)) {
+      auto &v = viewportVar.payload.float4Value;
+      // Read Float4 rect as (X0, Y0, X1, Y1)
+      viewItem.viewport = Rect::fromCorners(v[0], v[1], v[2], v[3]);
+    }
+
+    // (optional) set render target
+    SHVar rtVar{};
+    if (getFromTable(shContext, inputTable, "Target", rtVar)) {
+      SHRenderTarget *renderTarget = varAsObjectChecked<SHRenderTarget>(rtVar, Types::RenderTarget);
+      viewItem.renderTarget = renderTarget->renderTarget;
+    }
+
+    // (optional) Push window region for input
+    SHVar windowRegionVar{};
+    if (getFromTable(shContext, inputTable, "WindowRegion", windowRegionVar)) {
+      // Read Float4 rect as (X0, Y0, X1, Y1)
+      auto &v = windowRegionVar.payload.float4Value;
+      int4 region(v[0], v[1], v[2], v[3]);
+
+      // This is used for translating cursor inputs from window to view space
+      // TODO: Adjust relative to parent region, for now always assume this is in global window coordinates
+      inputItem.windowMapping = input::WindowSubRegion{
+          .region = region,
+      };
+
+      // Safe to automatically set reference size based on region
+      float2 regionSize = float2(region.z - region.x, region.w - region.y);
+      viewItem.referenceSize = int2(linalg::floor(float2(regionSize)));
+    }
+
+    auto &mwg = getMainWindowGlobals();
```

Comment:
do we only have mainwindow?

Reply:
Yes, but it's a member from the base consumer so you could have a separate wire with another window if you really wanted